### PR TITLE
De-duplicate BuildTarget.sourses

### PR DESCRIPTION
**Pull request by klokik**
If same source is provided by multiple dependencies it was added multiple times, as `added_sources` was only guarding against duplicates within the same source list. This was not a problem with ninja, but triggers multiple sanity checks within xcode backend while attempting to create multiple ids for the same file.

I encountered this problem with gtest_main from wrapdb.

### CODE REVIEW

File: mesonbuild/build.py [hunk 1]
```diff
@@ -758,6 +758,8 @@ def __init__(
         self.pch: T.Dict[str, T.List[str]] = {}
         self.extra_args: T.DefaultDict[str, T.List[str]] = kwargs.get('language_args', defaultdict(list))
         self.sources: T.List[File] = []
+        # If the same source is defined multiple times, use it only once.
+        self.added_sources: T.Set[File] = set()
```
Comment (eli-schwartz) [inline mesonbuild/build.py:762]: Can we take this opportunity to rename it to "seen_sources"?

**Review by eli-schwartz — APPROVED** [review]

File: mesonbuild/build.py (hunk 1 above)
Comment (klokik) [reply mesonbuild/build.py]: I have no strong opinion on that, so sure, why not.

**Comment by bruchar1** [comment]
There is a typo in commit message: s/sourses/sources/

**Comment by klokik** [comment]
Fixed typo

**Pull request merged**